framework/tlvf/src/src/WSC/m1.cpp

```diff
@@ -12,6 +12,46 @@
 
 using namespace WSC;
 
+std::shared_ptr<m1> m1::parse(ieee1905_1::tlvWsc &tlv)
+{
+    if (!tlv.payload_length()) {
+        TLVF_LOG(ERROR) << "No room to add attribute list (payload length = 0)";
+        return nullptr;
+    }
+    auto msg_type = wsc_msg_type_from_buffer(tlv.payload(), tlv.payload_length());
+    if (msg_type != eWscMessageType::WSC_MSG_TYPE_M1) {
+        TLVF_LOG(DEBUG) << "Not M1 message! message type " << int(msg_type);
+        return nullptr;
+    }
+    auto attributes = std::make_shared<m1>(tlv.payload(), tlv.payload_length(), true);
+    if (!attributes || !attributes->init()) {
+        TLVF_LOG(ERROR) << "Failed to initialize attributes";
+        return nullptr;
+    }
```

Comment:
IMO it would be much more natural to do the following:

```
auto attributes = std::make_shared<m1>(...);
attributes->init();
auto msg_type = attributes->getClass<cWscAttrMessageType>();
if (msg_type->msg_type != WSC_MSG_TYPE_M1) {
    LOG(INFO) << "Not M1, msg type is " << int(msg_type->msg_type);
    return nullptr;
}
```

That avoids the extra parsing of msg_type. Since we return nullptr, the invalid attrlist (which pretends to be an M1 but really is not) will be freed immediately, so we never expose an invalid M1 object.

Reply:
Agree, will be changed.